Terraform init command is failing to initilize AzureRm backend using Azure Devops (Terraform verion : 0.15.0)

Hello Team,

I am using Terraform to deploy microsoft azure resources using terraform in Azure DevOps.

Till yesterday day deployment was working fine and suddenly since today morning it started getting failed without any change.

Now what I understood is behind the scene azure devops is targetting terraform version 0.15.0 (latest) version which was 0.14.0 day before.

It is failing at terraform init command where it is trying to configure azurerm as a backend to manage terraform state.

Please find the following error logs for same.

**Note : I have replaced real values with <dummy_values> to maintain secrecy.**

2021-04-15T06:45:18.7341116Z ##[section]Starting: Terraform init
2021-04-15T06:45:18.7573359Z ==============================================================================
2021-04-15T06:45:18.7573782Z Task         : Terraform CLI
2021-04-15T06:45:18.7574247Z Description  : Execute terraform cli commands
2021-04-15T06:45:18.7574672Z Version      : 0.6.19
2021-04-15T06:45:18.7575005Z Author       : Charles Zipp
2021-04-15T06:45:18.7575654Z Help         : 
2021-04-15T06:45:18.7576760Z ==============================================================================
2021-04-15T06:45:19.7124290Z [command]C:\hostedtoolcache\windows\terraform\0.15.0\x64\terraform.exe version
2021-04-15T06:45:19.9539863Z Terraform v0.15.0
2021-04-15T06:45:19.9541562Z on windows_amd64
2021-04-15T06:45:19.9701294Z 
2021-04-15T06:45:19.9973463Z [command]C:\hostedtoolcache\windows\terraform\0.15.0\x64\terraform.exe init -backend-config=storage_account_name=<storage_account_name> -backend-config=container_name=<container_name> -backend-config=key=terraform.tfstate -backend-config=resource_group_name=<resource_group_name> -backend-config=arm_subscription_id=<arm_subscription_id> -backend-config=arm_tenant_id=*** -backend-config=arm_client_id=*** -backend-config=arm_client_secret=***
2021-04-15T06:45:20.2538384Z [0m[1mInitializing modules...[0m
2021-04-15T06:45:20.2553411Z - app_service in modules\app_service
2021-04-15T06:45:20.2564889Z - sqlserver_db in modules\sqlserver_db
2021-04-15T06:45:20.2573726Z - storage in modules\storage
2021-04-15T06:45:20.2621509Z 
2021-04-15T06:45:20.2622372Z [0m[1mInitializing the backend...[0m
2021-04-15T06:45:20.2644044Z [33m[33m╷[0m[0m
2021-04-15T06:45:20.2645949Z [33m│[0m [0m[1m[33mWarning: [0m[0m[1mVersion constraints inside provider configuration blocks are deprecated[0m
2021-04-15T06:45:20.2648133Z [33m│[0m [0m
2021-04-15T06:45:20.2649475Z [33m│[0m [0m[0m  on main.tf line 4, in provider "azurerm":
2021-04-15T06:45:20.2651392Z [33m│[0m [0m   4:   version = [4m"=2.20.0"[0m[0m
2021-04-15T06:45:20.2653042Z [33m│[0m [0m
2021-04-15T06:45:20.2655112Z [33m│[0m [0mTerraform 0.13 and earlier allowed provider version constraints inside the
2021-04-15T06:45:20.2656892Z [33m│[0m [0mprovider configuration block, but that is now deprecated and will be
2021-04-15T06:45:20.2658464Z [33m│[0m [0mremoved in a future version of Terraform. To silence this warning, move the
2021-04-15T06:45:20.2660177Z [33m│[0m [0mprovider version constraint into the required_providers block.
2021-04-15T06:45:20.2661230Z [33m╵[0m[0m
2021-04-15T06:45:20.2662206Z [0m[0m
2021-04-15T06:45:20.2663237Z [31m[31m╷[0m[0m
2021-04-15T06:45:20.2664545Z [31m│[0m [0m[1m[31mError: [0m[0m[1mInvalid backend configuration argument[0m
2021-04-15T06:45:20.2665823Z [31m│[0m [0m
2021-04-15T06:45:20.2667116Z [31m│[0m [0m[0mThe backend configuration argument "arm_subscription_id" given on the
2021-04-15T06:45:20.2668555Z [31m│[0m [0mcommand line is not expected for the selected backend type.
2021-04-15T06:45:20.2669708Z [31m╵[0m[0m
2021-04-15T06:45:20.2670702Z [0m[0m
2021-04-15T06:45:20.2671987Z [31m[31m╷[0m[0m
2021-04-15T06:45:20.2673315Z [31m│[0m [0m[1m[31mError: [0m[0m[1mInvalid backend configuration argument[0m
2021-04-15T06:45:20.2674403Z [31m│[0m [0m
2021-04-15T06:45:20.2675999Z [31m│[0m [0m[0mThe backend configuration argument "arm_tenant_id" given on the command
2021-04-15T06:45:20.2677337Z [31m│[0m [0mline is not expected for the selected backend type.
2021-04-15T06:45:20.2678488Z [31m╵[0m[0m
2021-04-15T06:45:20.2679529Z [0m[0m
2021-04-15T06:45:20.2680596Z [31m[31m╷[0m[0m
2021-04-15T06:45:20.2681908Z [31m│[0m [0m[1m[31mError: [0m[0m[1mInvalid backend configuration argument[0m
2021-04-15T06:45:20.2683002Z [31m│[0m [0m
2021-04-15T06:45:20.2684317Z [31m│[0m [0m[0mThe backend configuration argument "arm_client_id" given on the command
2021-04-15T06:45:20.2685589Z [31m│[0m [0mline is not expected for the selected backend type.
2021-04-15T06:45:20.2689517Z [31m╵[0m[0m
2021-04-15T06:45:20.2690483Z [0m[0m
2021-04-15T06:45:20.2691526Z [31m[31m╷[0m[0m
2021-04-15T06:45:20.2692829Z [31m│[0m [0m[1m[31mError: [0m[0m[1mInvalid backend configuration argument[0m
2021-04-15T06:45:20.2694101Z [31m│[0m [0m
2021-04-15T06:45:20.2695680Z [31m│[0m [0m[0mThe backend configuration argument "arm_client_secret" given on the command
2021-04-15T06:45:20.2711862Z [31m│[0m [0mline is not expected for the selected backend type.
2021-04-15T06:45:20.2712711Z [31m╵[0m[0m
2021-04-15T06:45:20.2713445Z [0m[0m
2021-04-15T06:45:20.2846527Z 
2021-04-15T06:45:20.3026280Z ##[error]Terraform command 'init' failed with exit code '1'.
2021-04-15T06:45:20.3092716Z ##[error][31m[31m╷[0m[0m
[31m│[0m [0m[1m[31mError: [0m[0m[1mInvalid backend configuration argument[0m
[31m│[0m [0m
[31m│[0m [0m[0mThe backend configuration argument "arm_subscription_id" given on the
[31m│[0m [0mcommand line is not expected for the selected backend type.
[31m╵[0m[0m
[0m[0m
[31m[31m╷[0m[0m
[31m│[0m [0m[1m[31mError: [0m[0m[1mInvalid backend configuration argument[0m
[31m│[0m [0m
[31m│[0m [0m[0mThe backend configuration argument "arm_tenant_id" given on the command
[31m│[0m [0mline is not expected for the selected backend type.
[31m╵[0m[0m
[0m[0m
[31m[31m╷[0m[0m
[31m│[0m [0m[1m[31mError: [0m[0m[1mInvalid backend configuration argument[0m
[31m│[0m [0m
[31m│[0m [0m[0mThe backend configuration argument "arm_client_id" given on the command
[31m│[0m [0mline is not expected for the selected backend type.
[31m╵[0m[0m
[0m[0m
[31m[31m╷[0m[0m
[31m│[0m [0m[1m[31mError: [0m[0m[1mInvalid backend configuration argument[0m
[31m│[0m [0m
[31m│[0m [0m[0mThe backend configuration argument "arm_client_secret" given on the command
[31m│[0m [0mline is not expected for the selected backend type.
[31m╵[0m[0m
[0m[0m

2021-04-15T06:45:20.7056806Z ##[section]Finishing: Terraform init

--- a/website/upgrade-guides/0-15.html.markdown
+++ b/website/upgrade-guides/0-15.html.markdown
@@ -52,6 +52,7 @@ Upgrade guide sections:
 * [Commands Accepting a Configuration Directory Argument](#commands-accepting-a-configuration-directory-argument)
 * [Microsoft Windows Terminal Support](#microsoft-windows-terminal-support)
 * [Other Minor Command Line Behavior Changes](#other-minor-command-line-behavior-changes)
+* [Azure Backend `arm_`-prefixed Arguments](#azure-backend-removed-arguments)
 
 ## Sensitive Output Values
 
@@ -493,3 +494,44 @@ cleanup of obsolete features and improved consistency:
 
     If you are using `-force` in an automated call to `terraform destroy`,
     change to using `-auto-approve` instead.
+
+## Azure Backend Removed Arguments
+
+In an earlier release the `azure` backend changed to remove the `arm_` prefix
+from a number of the configuration arguments:
+
+| Old Name              | New Name          |
+|-----------------------|-------------------|
+| `arm_client_id`       | `client_id`       |
+| `arm_client_secret`   | `client_secret`   |
+| `arm_subscription_id` | `subscription_id` |
+| `arm_tenant_id`       | `tenant_id`       |
+
+The old names were previously deprecated, but we've removed them altogether
+in Terraform v0.15 in order to conclude that deprecation cycle.
+
+If you have a backend configuration using the old names then you may see
+errors like the following when upgrading to Terraform v0.15:
+
+```
+╷
+│ Error: Invalid backend configuration argument
+│
+│ The backend configuration argument "arm_client_id" given on
+│ the command line is not expected for the selected backend type.
+╵
+```
+
+If you see errors like this, rename the arguments in your backend configuration
+as shown in the table above and then run the following to re-initialize your
+backend configuration:
+
+```
+terraform init -reconfigure
+```
+
+The `-reconfigure` argument instructs Terraform to just replace the old
+configuration with the new configuration directly, rather than offering to
+migrate the latest state snapshots from the old to the new configuration.
+Migration would not be appropriate in this case because the old and new
+configurations are equivalent and refer to the same remote objects.

